### PR TITLE
New Intervention command

### PR DESCRIPTION
**Pull request by ibreschi**
## Make an inference within one Intervention
```
deepo site intervention
error: the following arguments are required: intervention command
usage: deepo site intervention [-h] [--verbose]
                               {create,delete,inference,status} ...
positional arguments:
  {create,delete,inference,status}
    create              Create a new intervention
    delete              Delete a new intervention
    infer               Make an inference on an intervention
    status              Retrieve the intervention status
optional arguments:
  -h, --help            show this help message and exit
  --verbose             Increase output verbosity.
```

## Create an Intervention
```
deepo site intervention create -n "ciao" --base_url $base_url 
```

## Get Intervention state
```
deepo site intervention status -i $intervention_id  --base_url $base_url 
```

## Delete an Intervention
```
deepo site intervention delete -i  $intervention_id --base_url $base_url  
```
## Make an inference within one Intervention
```
deepo site intervention infer --base_url $base_url --intervention_id $intervention_id -e windshield_image@image@https://images.com/windshield_image.jpg closeup_image@image@https://images.com/closeup_image.jpg car_name@text@fiat
```

### CODE REVIEW

File: deepomatic/cli/lib/site.py [hunk 1]
```diff
@@ -160,3 +167,42 @@ def _run_docker_compose(self, *args):
...
+
+    def create_intervention(self, base_url, name, metadata):
+        intervention_url = f"{base_url}/interventions"
+        data = {'name': name,
+                'metadata': metadata
+                }
+
+        res = requests.post(f'{intervention_url}/', data=json.dumps(data), headers=headers)
+        if res.status_code == 201:
+            return res.json()['intervention_id']
```
Comment (maingoh) [inline deepomatic/cli/lib/site.py:179]: should be `id` now

File: deepomatic/cli/cmds/site/intervention/create.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        parser.add_argument('-n', '--name', required=True, type=str, help="Intervention name")
+        parser.add_argument('-m', "--metadata", dest='metadata',
+                            action=BuildDict, nargs="+", metavar="NAME:VAL",
+                            help='Metadata for the intervention in format name:value')
+        return parser
+
+    def run(self, base_url, name, metadata, **kwargs):
+        metadata = {'version': '1.1',
+                    # NOTE: Temporary field to pass the question to the CustomerAPI
+                    'question_ids': ['1c27cba0-5fb7-4417-9fd1-683bc70df8ce']}
```
Comment (Zanoellia) [inline deepomatic/cli/cmds/site/intervention/create.py:22]:
```suggestion
        metadata = {'version': '1.1'}
```
Question id not necessary anymore

File: deepomatic/cli/cmds/site/intervention/inference.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        parser = super(InferCommand, self).setup(subparsers)
+        parser.add_argument('-u', "--base_url", required=True, type=str, help="url of your api")
+        parser.add_argument('-i', '--intervention_id', required=True, type=str, help="Intervention id")
+        parser.add_argument('-e', "--entries", dest='entries',
+                            action=BuildEntries, nargs="+", metavar="NAME@TYPE@VAL",
+                            help='Entries for the workflow in format entry_name@entry_type@value',
+                            required=True)
+        parser.add_argument('-m', "--metadata", dest='metadata',
+                            action=BuildDict, nargs="+", metavar="NAME:VAL",
+                            help='Metadata for the inference in format name:value')
```
Comment (Zanoellia) [inline deepomatic/cli/cmds/site/intervention/inference.py:84]: Maybe add `required=True` as we need at least one metadata for now?

File: deepomatic/cli/cmds/site/intervention/create.py (hunk 2 above)
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/create.py:22]: yes ! a relic of the past :D

File: deepomatic/cli/cmds/site/intervention/inference.py (hunk 3 above)
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/inference.py:84]: yes!

File: deepomatic/cli/lib/site.py [hunk 4]
```diff
@@ -44,6 +39,10 @@ def __init__(self, path=DEEPOMATIC_SITE_PATH, client_cls=HTTPHelper):
         makedirs(path)
         self._repo = Repo.init(path)
         self._client = client_cls()
+        test_api_key = os.getenv('DEEPOMATIC_API_KEY')
```
Comment (Zanoellia) [inline deepomatic/cli/lib/site.py:42]: Why not `api_key` only?

File: deepomatic/cli/cmds/site/intervention/status.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class StatusCommand(Command):
+    """
+        Retrieve the intervention status
+    """
+
+    def setup(self, subparsers):
+        parser = super(StatusCommand, self).setup(subparsers)
+        parser.add_argument('-u', "--base_url", required=True, type=str, help="url of your api")
```
Comment (Zanoellia) [inline deepomatic/cli/cmds/site/intervention/status.py:12]: Maybe precise that the api url is the customer api one?

File: deepomatic/cli/cmds/site/intervention/inference.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    elif type == "number":
+        return value
+    else:
+        raise Exception("type {} not supported".format(type))
+
+
+class BuildEntries(argparse.Action):
+    """
+    This class is used in argparse. It will transform a chain of name@type@values into a list.
```
Comment (Zanoellia) [inline deepomatic/cli/cmds/site/intervention/inference.py:32]: nit: tell that the format of objects in the built list are customer-api compatible

File: deepomatic/cli/lib/site.py [hunk 7]
```diff
@@ -173,3 +179,42 @@ def _run_docker_compose(self, *args):
...
+        else:
+            return res.text
+
+    def status_intervention(self, base_url, intervention_id):
+        intervention_url = "{}/interventions".format(base_url)
+        res = requests.get('{}/{}'.format(intervention_url, intervention_id), headers=self.headers)
+        if res.status_code == 200:
+            return res.json()
+        else:
+            return res.text
```
Comment (maingoh) [inline deepomatic/cli/lib/site.py:201]: technically deepocli should exit with a bad code if it goes there

File: tests/test_site.py [hunk 8]
```diff
@@ -269,3 +271,19 @@ def test_site_deployment_manifest(self):
             assert '- name: workflow-server' in message
             assert '- name: customer-api' in message
             assert 'kind: Ingress' in message
+
+    def test_intervention(self):
+        bouyges_stag = 'https://api.80fa58aa-c834-4a3f-87b9.stag.customers.deepomatic.com'
```
Comment (maingoh) [inline tests/test_site.py:276]: can you add this by env var ? and rename this var

File: deepomatic/cli/cmds/site/fetch.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+from ..utils import Command
+from ...lib.site import SiteManager
+
+
+class FetchCommand(Command):
+    """
+        Fetch a site
```
Comment (maingoh) [inline deepomatic/cli/cmds/site/fetch.py:7]: Not very needed I think

File: tests/test_site.py [hunk 10]
```diff
@@ -273,15 +275,15 @@ def test_site_deployment_manifest(self):
...
         intervention_id = result
         assert len(intervention_id) == 36
 
         args = "site intervention status -i {} --base_url {}".format(intervention_id, bouyges_stag)
-        result = call_deepo(args)
-        assert list(result.keys()) == ['id', 'name', 'config_id', 'questions', 'answers', 'site_id',
-                                       'app_version_id', 'start_date', 'end_date', 'update_date',
-                                       'review_date', 'tags', 'assigned_user_id', 'enabled', 'metadata']
+        result = call_deepo(args, api_key=customer_api_api_key)
+        assert set(result.keys()) == set(['id', 'name', 'config_id', 'questions', 'answers', 'site_id',
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply tests/test_site.py:284]: For some esoteric mystery of python compatibility, 2.7 and 3.x have different behaviors in comparing lists. when using a set everybody is happy

[the remark this answers is not in the document]
Comment (Zanoellia) [reply tests/test_site.py:284]: haha sad but true :see_no_evil:

File: deepomatic/cli/cmds/site/fetch.py (hunk 9 above)
Comment (ibreschi) [reply deepomatic/cli/cmds/site/fetch.py:7]: removed

File: deepomatic/cli/cmds/site/fetch.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+from ..utils import Command
+from ...lib.site import SiteManager
+
+
+class FetchCommand(Command):
```
Comment (maingoh) [inline deepomatic/cli/cmds/site/fetch.py:5]: Maybe add a test

File: deepomatic/cli/cmds/site/intervention/delete.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
+from ...utils import Command
+from ..utils import SiteManager
+
+
+class DeleteCommand(Command):
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/delete.py:5]: you can delete one intervention in the customer-api

File: deepomatic/cli/cmds/site/intervention/inference.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+    def setup(self, subparsers):
+        parser = super(InferCommand, self).setup(subparsers)
+        parser.add_argument('-u', "--base_url", required=True, type=str, help="url of your api")
+        parser.add_argument('-i', '--intervention_id', required=True, type=str, help="Intervention id")
+        parser.add_argument('-e', "--entries", dest='entries',
+                            action=BuildEntries, nargs="+", metavar="NAME@TYPE@VAL",
+                            help='Entries for the workflow in format entry_name@entry_type@value',
+                            required=True)
+        parser.add_argument('-m', "--metadata", dest='metadata', required=True,
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/inference.py:73]: -m Make_The_WF_Play_Prog:True

File: deepomatic/cli/cmds/site/intervention/inference.py (hunk 6 above)
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/inference.py:32]: done

File: deepomatic/cli/cmds/site/intervention/status.py (hunk 5 above)
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/status.py:12]: done

File: deepomatic/cli/lib/site.py (hunk 4 above)
Comment (ibreschi) [reply deepomatic/cli/lib/site.py:42]: done

File: tests/test_site.py (hunk 8 above)
Comment (ibreschi) [reply tests/test_site.py:276]: done

File: deepomatic/cli/lib/site.py [hunk 14]
```diff
@@ -173,3 +179,42 @@ def _run_docker_compose(self, *args):
             except OSError:
                 pass
             p.wait()
+
+    def create_intervention(self, base_url, name, metadata):
+        intervention_url = "{}/interventions".format(base_url)
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/lib/site.py:184]: done

File: deepomatic/cli/cmds/site/intervention/delete.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (maingoh) [reply deepomatic/cli/cmds/site/intervention/delete.py:5]: Oh, I thought it wasn't implemented

File: deepomatic/cli/lib/site.py [hunk 15]
```diff
@@ -173,3 +179,42 @@ def _run_docker_compose(self, *args):
...
+        data = {'name': name,
+                'metadata': metadata
+                }
+
+        res = requests.post('{}/'.format(intervention_url), data=json.dumps(data), headers=self.headers)
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/lib/site.py:189]: `{"detail":"JSON parse error - Expecting value: line 1 column 1 (char 0)"}`

File: deepomatic/cli/cmds/site/intervention/inference.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (maingoh) [reply deepomatic/cli/cmds/site/intervention/inference.py:73]: but it is unused :smile:

File: deepomatic/cli/lib/site.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (maingoh) [reply deepomatic/cli/lib/site.py:189]: Hmm, weird, anyway leave it like this

File: deepomatic/cli/cmds/site/intervention/inference.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Zanoellia) [reply deepomatic/cli/cmds/site/intervention/inference.py:73]: It's mandatory for the customer-api so we'll need to modify it first.

File: deepomatic/cli/cmds/site/intervention/inference.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+import argparse
+import base64
+
+
+def build_data(type, value):
+    if type == "text":
+        return value
+    elif type == "image":
+        # Super naive test
+        if 'http' in value:
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/inference.py:13]: done

File: deepomatic/cli/cmds/site/intervention/delete.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ibreschi) [reply deepomatic/cli/cmds/site/intervention/delete.py:5]: the customer api reserves plenty of surprises ^^

File: deepomatic/cli/cmds/site/intervention/inference.py (hunk 6 above)
Comment (Zanoellia) [reply deepomatic/cli/cmds/site/intervention/inference.py]: Nice the class renaming!

File: deepomatic/cli/cmds/site/intervention/inference.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    This class is used in argparse. It will transform a chain of name@type@values into a list.
+    """
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        self._nargs = nargs
+        super(BuildEntries, self).__init__(option_strings, dest, nargs=nargs, **kwargs)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        entries = []
+        for kv in values:
+            vals = kv.split("@")
```
[the remark this answers is not in the document]
Comment (Zanoellia) [reply deepomatic/cli/cmds/site/intervention/inference.py:41]: Wait for the day we'll need to have predictions or regions as entries :joy_cat:

**Review by maingoh — APPROVED** [review]
Perfect !

File: deepomatic/cli/lib/site.py [hunk 18]
```diff
@@ -173,3 +180,46 @@ def _run_docker_compose(self, *args):
...
+    def make_intervention_url(self, base_url):
+        return "{}/interventions".format(base_url)
+
+    def create_intervention(self, base_url, name, metadata):
+        intervention_url = self.make_intervention_url(base_url)
+        data = {'name': name,
+                'metadata': metadata
+                }
+
+        res = self.session.post('{}/'.format(intervention_url), data=json.dumps(data))
```
Comment (maingoh) [inline deepomatic/cli/lib/site.py:193]:
```suggestion
        res = self.session.post(intervention_url, data=json.dumps(data))
```

File: deepomatic/cli/lib/site.py [hunk 19]
```diff
@@ -173,3 +180,46 @@ def _run_docker_compose(self, *args):
...
+        else:
+            return res.text
+
+    def make_inference(self, base_url, intervention_id, entries, metadata):
+        intervention_url = self.make_intervention_url(base_url)
+        input_data_url = "{}/{}/input_data".format(intervention_url, intervention_id)
+        data = {'inputs': entries,
+                "metadata": metadata}
+
+        res = self.session.post('{}/'.format(input_data_url), data=json.dumps(data))
```
Comment (maingoh) [inline deepomatic/cli/lib/site.py:221]:
```suggestion
        res = self.session.post(input_data_url, data=json.dumps(data))
```

File: deepomatic/cli/lib/site.py (hunk 18 above)
Comment (ibreschi) [reply deepomatic/cli/lib/site.py:193]: done

File: deepomatic/cli/lib/site.py (hunk 19 above)
Comment (ibreschi) [reply deepomatic/cli/lib/site.py:221]: done

**Review by Alexis-Jacob — APPROVED** [review]

**Review by Zanoellia — APPROVED** [review]
Well done 👏

**Pull request merged**